site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md

```diff
@@ -0,0 +1,182 @@
+	if err != nil {
+		log.Fatalf("Unable to parse %s folder", folder)
+	}
+	// 2. find the required package by name
+	var appPkg *ast.Package
+	for _, pkg := range pkgs {
+		if pkg.Name == pkgName {
+			appPkg = pkg
+			break
+		}
+	}
+	if appPkg == nil {
+		log.Fatalf("Unable to find package %s", pkgName)
+	}
+
+	// 3. find all methods that are bound to the specific struct
+	for _, file := range appPkg.Files {
+		ast.Inspect(file, func(n ast.Node) bool {
+			if fun, ok := n.(*ast.FuncDecl); ok {
+				// 4. Validate that method is exported and has a receiver
+				if fun.Name.IsExported() && fun.Recv != nil && len(fun.Recv.List) == 1 {
+						// 5. Check that the receiver is actually the struct we want
+						if r, rok := fun.Recv.List[0].Type.(*ast.StarExpr); rok && r.X.(*ast.Ident).Name == structName {
+							// we found it!
+						}
+					}
+				}
+
+			}
+			return true
+		})
+	}
+```
+
+Steps 1 and 2 are pretty straightforward, the interesting bits start at step 3: For each file in the package, we execute `ast.Inspect` to get all the AST nodes. For every node that is actually a function (checked by `n.(*ast.FuncDecl)`), we check if that function is:  
+* Exported (we are not interested in private methods)
+* Has a receiver (it's bound to a struct)
+* Receiver's type matches the struct we are interested in
+
+### Collecting functions and their comments
+
+Now that we can get a `*ast.FuncDecl` (let's call it `fun`) for each function, we can collect all the information about it and reconstruct it:
+
+1. Name: `fun.Name.Name`
+2. Comments: `fun.Doc.List`
+3. Parameters: fun.Type.Params.Lis
+4. Return values: fun.Type.Results.List
+
+### Generate the interface file
+
+To generate the output interface, we'll use Go's `template` package. Let's define a simple template:
+
+```go
+const outputTemplate = `
+// DO NOT EDIT, auto generated
+
+package {{.Package}}
+
+type {{.Name}} interface {
+	{{.Content}}
+}
+`
+```
+
+And populate it:
+
+```go
+	sort.Strings(funcs)
+	out := bytes.NewBufferString("")
+
+	t := template.Must(template.New("").Parse(outputTemplate))
+	err = t.Execute(out, map[string]interface{}{
+		"Content": strings.Join(funcs, "\n"),
+		"Name":    ifName,
+		"Package": pkgName,
+	})
+```	
+
+We are almost done! Our interface file is ready, but it's missing a crucial part - imports. Luckily, there is a package for that™ - https://pkg.go.dev/golang.org/x/tools/imports
+
+Let give it whirl:
+```go
+	formatted, err := imports.Process(outputFile, out.Bytes(), &imports.Options{Comments: true})
+	if err != nil {
+		log.Panic(err)
+	}
+	err = ioutil.WriteFile(outputFile, formatted, 0644)
+```	
+
+Voila! You have an interface file that contains all the methods implemented on the `struct`.
+
+## struct2interface
+
+The process I've described is a rather generic one, so to make it fully repeatable, I've extracted it into a separate CLI utility called `struct2interface`. It can be found at https://github.com/reflog/struct2interface
```

Comment:
```suggestion
The process I've described is a rather generic one, so to make it fully repeatable, I've extracted it into a separate CLI utility called `struct2interface`. It can be found at https://github.com/reflog/struct2interface.
```